File: Leetcode/Dynamic Programming/647. Palindromic Substrings.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int N = s.size(), palindromes = 0;
        int isPalindrome[N][N];
        memset(isPalindrome, 0, sizeof(isPalindrome));
        for (int length = 1; length <= N; length++) {
            int i = 0, j = i + length - 1;
            for (int i = 0, j = i + length - 1; j < N; i++, j++) {
                if (length == 1) isPalindrome[i][j] = 1;
                else if (length == 2) isPalindrome[i][j] = s[i] == s[j];
                else isPalindrome[i][j] = (s[i] == s[j] && isPalindrome[i + 1][j - 1]);
                palindromes += isPalindrome[i][j];
            }
        }
        return palindromes;
    }
};

File: Leetcode/Dynamic Programming/647. Palindromic Substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        N = len(s)
        palindromes = 0
        isPalindrome = [[0] * N for _ in range(N)]
        for length in range(1, N + 1):
            i = 0
            for j in range(i + length - 1, N):
                if (length == 1) or (length == 2 and s[i] == s[j]):
                    isPalindrome[i][j] = 1
                else:
                    isPalindrome[i][j] = 1 if s[i] == s[j] and isPalindrome[i + 1][j - 1] else 0
                palindromes += isPalindrome[i][j]
                i += 1
        return palindromes
